### PR TITLE
OBB frustum culling algorithm and unit tests

### PR DESCRIPTION
**Pull request by otsakir**
Oriented bounding box (OBB) - frustum intersection algorithm

### CODE REVIEW

File: engine/dlib/src/dlib/intersection.cpp [hunk 1]
```diff
@@ -98,4 +98,52 @@ bool TestFrustumSphere(const Frustum& frustum, const dmVMath::Vector4& pos, floa
     return true;
 }
 
+// Returns 'false' if the bounding box is off the frustum. Returning 'true' does not guarantee that the object intersects the frustum or is inside it.
+bool TestFrustumOBB(const Frustum& frustum, const dmVMath::Matrix4& world, dmVMath::Vector3& aabb_min, dmVMath::Vector3& aabb_max)
```
Comment (JCash) [inline engine/dlib/src/dlib/intersection.cpp:102]: As an optimisation, I also want you to support the "skip_near_far" boolean that you see in the other functions.

This allows us to skip planes that one sometimes doesn't care about.

File: engine/dlib/src/dlib/intersection.cpp [hunk 2]
```diff
@@ -98,4 +98,52 @@ bool TestFrustumSphere(const Frustum& frustum, const dmVMath::Vector4& pos, floa
...
+
+    // for any of the six frustum planes if the all corner points lie in the negative halfspace, do cull the object
+    for (int plane_i=0; plane_i<6; plane_i++)
+    {
+        // get plane normal/equation
+        dmVMath::Vector4 plane_normal = frustum.m_Planes[plane_i];
+        bool positive_found = false;
+        for (int corner_i=1; corner_i<8; corner_i++)
+        {
+            float dotproduct = Vectormath::Aos::dot(corner_points[corner_i], plane_normal);
```
Comment (JCash) [inline engine/dlib/src/dlib/intersection.cpp:133]: This dot product seems off.

The point is in world space, and since the normal is just a normal, what you'll get is the distance from origin, not the distance to the plane.

See the usage of `DistanceToPlane` for reference.

File: engine/dlib/src/dlib/intersection.cpp [hunk 3]
```diff
@@ -98,4 +98,52 @@ bool TestFrustumSphere(const Frustum& frustum, const dmVMath::Vector4& pos, floa
...
+            if (dotproduct >= 0)
```
Comment (JCash) [inline engine/dlib/src/dlib/intersection.cpp:134]: As I mentioned earlier, doing per-point checks leads to trouble.

We should use a SAT test (Separating axis theorem).
I.e. find the min/max extents (call them d1 and d2) of the box when projected onto the planes normal.
If:
* If the the sign of the two values are differing (`d1 * d2 < 0.0f`), then there is for sure an intersection
* If both d1 and d2 are positive, then the OBB is in front of the plane


In short: For each point in the OOB, we at the same time measure the distance to the plane for its _opposite_ point.



Instead, use the _range_ that

File: engine/dlib/src/dlib/intersection.cpp [hunk 4]
```diff
@@ -98,4 +98,52 @@ bool TestFrustumSphere(const Frustum& frustum, const dmVMath::Vector4& pos, floa
...
+                break; // no need to check for the rest of the points
+            }
+        }
+        if (! positive_found) // if all corners are in the negative halfspace for the plane, we're done
+        {
+            return false; // no intersection, do cull
+        }
+
+    }
+    return true; // probalby inside the frustum but not 100% sure
```
Comment (JCash) [inline engine/dlib/src/dlib/intersection.cpp:146]: "Probalby"

The comment is also interesting, since you don't elaborate on why it isn't 100% sure?

File: engine/dlib/src/test/test_intersection.cpp [hunk 5]
```diff
@@ -105,6 +105,61 @@ TEST(dmVMath, TestFrustumSphere)
 
 }
 
+TEST(dmVMath, TestFrustumOBB)
+{
+    dmVMath::Matrix4 proj = dmVMath::Matrix4::orthographic(0.0f, FRUSTUM_WIDTH, 0.0f, FRUSTUM_HEIGHT, FRUSTUM_NEAR, FRUSTUM_FAR);
```
Comment (JCash) [inline engine/dlib/src/test/test_intersection.cpp:110]: I would also like to see tests with a non orthographics projection matrix, as that's what most users will use with this function.

You can ofc choose constants that make it easier to reason about in the tests.

**Review by JCash — CHANGES_REQUESTED** [review]
Overall I think there are a lot of calculations going on and I feel it should be possible to also simplify it a bit.

E.g. what if we for each plane in the frustum transform the plane into the space of the OOBB. Then we make it into a problem of AABB vs Plane intersections instead.
https://gdbooks.gitbooks.io/3dcollisions/content/Chapter2/static_aabb_plane.html

E:g. this approach to transform a plane using a matrix4 seems promising:
https://stackoverflow.com/a/48408289/468516

File: engine/dlib/src/dlib/intersection.cpp (hunk 2 above)
Comment (JCash) [reply engine/dlib/src/dlib/intersection.cpp:133]: Ah, good call, I missed the fact that the W argument is 1.
That's what the Point3 is for, and I got confused by the Vector4.

File: engine/dlib/src/dlib/intersection.cpp [hunk 6]
```diff
@@ -98,4 +98,52 @@ bool TestFrustumSphere(const Frustum& frustum, const dmVMath::Vector4& pos, floa
...
+        // get plane normal/equation
+        dmVMath::Vector4 plane_normal = frustum.m_Planes[plane_i];
+        bool positive_found = false;
+        for (int corner_i=1; corner_i<8; corner_i++)
+        {
+            float dotproduct = Vectormath::Aos::dot(corner_points[corner_i], plane_normal);
+            if (dotproduct >= 0)
+            {
+                positive_found = true;
+                break; // no need to check for the rest of the points
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/dlib/src/dlib/intersection.cpp:137]: NVM, I was under the impression that I had made the frustum normals point outwards.

File: engine/dlib/src/dlib/intersection.cpp (hunk 4 above)
Comment (JCash) [reply engine/dlib/src/dlib/intersection.cpp:146]: I still think the comment is "wrong", in that it gives the impression we are not sure what we are doing.
Instead, you could state what false positives _are_ allowed to happen.

File: engine/dlib/src/dlib/intersection.cpp (hunk 2 above)
Comment (JCash) [inline engine/dlib/src/dlib/intersection.cpp:133]: This is a good improvement to `DistanceToPlane(Plane plane, dmVMath::Vector4 pos)`.
Please update that function and use that here instead.

File: engine/dlib/src/dlib/intersection.cpp (hunk 4 above)
Comment (otsakir) [reply engine/dlib/src/dlib/intersection.cpp:146]: yes, It's kind of funny to read. And to write. I pondered a little on the return type (bool) and the name of the function which prepares the reader for a clean and strict piece of functionality.

Ok, I'll update the comment accordingly.

File: engine/dlib/src/dlib/intersection.cpp [hunk 7]
```diff
@@ -98,4 +98,52 @@ bool TestFrustumSphere(const Frustum& frustum, const dmVMath::Vector4& pos, floa
...
+            }
+        }
+        if (! positive_found) // if all corners are in the negative halfspace for the plane, we're done
+        {
+            return false; // no intersection, do cull
+        }
```
[the remark this answers is not in the document]
Comment (otsakir) [reply engine/dlib/src/dlib/intersection.cpp:143]: I'm not sure if i have to answer this or it's already settled.

If **all** corner points are in the negative halfspace of **any** one frustum plane (i.e. !positive_found) we don't care if they in front of another plane or not. They are for sure outside the frustum volume, so we can safely return 'false'.

File: engine/dlib/src/test/test_intersection.cpp (hunk 5 above)
Comment (otsakir) [reply engine/dlib/src/test/test_intersection.cpp:110]: Ok, i'll add another test for perspective projection. 

I'm not sure wdym by "constants" though.

Comment (otsakir) [reply engine/dlib/src/test/test_intersection.cpp:125]: Done.

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/dlib/src/test/test_intersection.cpp [hunk 8]
```diff
@@ -113,80 +128,111 @@ TEST(dmVMath, TestFrustumOBB)
...
+    float BOX_SIDE = 1;
     dmVMath::Vector3 minPoint(0,0,0);
     dmVMath::Vector3 maxPoint(BOX_SIDE,BOX_SIDE,BOX_SIDE); // note, the box lies in the +Z axis
 
     dmVMath::Matrix4 trans;
 
     // place BB at (0,0,0), outside the frustum that starts at -FRUSTUM_NEAR
     trans = dmVMath::Matrix4::identity();
-    ASSERT_FALSE(dmIntersection::TestFrustumOBB(frustum, trans, minPoint, maxPoint));
+    ASSERT_FALSE(dmIntersection::TestFrustumOBB(frustum, trans, minPoint, maxPoint, false));
```
[the remark this answers is not in the document]
Comment (otsakir) [reply engine/dlib/src/test/test_intersection.cpp:139]: yup, makes sense.

File: engine/dlib/src/dlib/intersection.cpp [hunk 9]
```diff
@@ -123,14 +122,14 @@ bool TestFrustumOBB(const Frustum& frustum, const dmVMath::Matrix4& world, dmVMa
...
+    int max_planes = skip_near_far ? 4 : 6;
+    for (int plane_i=0; plane_i<max_planes; plane_i++)
     {
         // get plane normal/equation
-        dmVMath::Vector4 plane_normal = frustum.m_Planes[plane_i];
         bool positive_found = false;
         for (int corner_i=1; corner_i<8; corner_i++)
         {
-            float dotproduct = Vectormath::Aos::dot(corner_points[corner_i], plane_normal);
+            float dotproduct = DistanceToPlane(frustum.m_Planes[plane_i], corner_points[corner_i]);
```
[the remark this answers is not in the document]
Comment (otsakir) [reply engine/dlib/src/dlib/intersection.cpp:132]: ok

File: engine/dlib/src/test/test_intersection.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (otsakir) [reply engine/dlib/src/test/test_intersection.cpp:138]: done

**Review by JCash — APPROVED** [review]
